python/lsst/pipe/base/shims.py

```diff
@@ -0,0 +1,320 @@
+            A Generation 3 data ID that identifies the dataset.
+        doBackup : `bool`
+            This option is provided for compatibility with
+            `lsst.daf.persistence.Butler`, but must be `False`.
+        rest
+            Additional key-value pairs to augment the given data ID.
+        """
+        if doBackup:
+            raise NotImplementedError("ShimButler cannot implement put with 'doBackup=True'")
+        datasetType = self._translateDatasetType(datasetType)
+        self._butler3.put(obj, datasetType, self._makeDataId(dataId, **rest))
+
+    def dataRef(self, datasetType, level=None, dataId=None, **rest):
+        """Return a DataRef associated with the given dataset type and data ID.
+
+        Parameters
+        ----------
+        datasetType : `str`
+            Name of the dataset type.
+        dataId : `dict` or `~lsst.daf.butler.DataId`, optional
+            A Generation 3 data ID that identifies the dataset.
+        level
+            This option is provided for compatibility with
+            `lsst.daf.persistence.Butler`, but must be `None`.
+        rest
+            Additional key-value pairs to augment the given data ID.
+        """
+        if level is not None:
+            raise NotImplementedError("ShimButler cannot implement dataRef with 'level != None'")
+        fullDataId = {}
+        if dataId is not None:
+            fullDataId.update(dataId)
+        fullDataId.update(rest)
+        return next(iter(ShimButlerSubset(self, datasetType, [fullDataId])))
+
+
+class ShimButlerSubset:
+    """A shim for a Gen2 `~lsst.daf.persistence.ButlerSubset` with a Gen3
+    `~lsst.daf.butler.Butler` backend.
+
+    Parameters
+    ----------
+    butler : `ShimButler`
+        Butler shim instance.
+    datasetType : `str`
+        Name of the dataset type.
+    dataIds : iterable of `dict` or `~lsst.daf.butler.DataId`
+        Generation 3 data IDs that define the data in this subset.
+    """
+
+    GENERATION = 2.5
+    """This is a Generation 2 shim for a Generation3 Butler.
+    """
+
+    def __init__(self, butler, datasetType, dataIds):
+        self.butler = butler
+        self.datasetType = datasetType
+        self._dataIds = list(dataIds)
+
+    def __len__(self):
+        return len(self._dataIds)
+
+    def __iter__(self):
+        for dataId in self._dataIds:
+            yield ShimDataRef(self, dataId)
+
+
+class ShimDataRef:
+    """A shim for a Gen2 `~lsst.daf.persistence.ButlerDataRef` with a Gen3
+    `~lsst.daf.butler.Butler` backend.
+
+    Parameters
+    ----------
+    butlerSubset : `ShimButlerSubset`
+        ButlerSubset shim instance.
+        Name of the dataset type.
+    dataId : `dict` or `~lsst.daf.butler.DataId`
+        Generation 3 data ID associated with this reference.
+    """
+
+    GENERATION = 2.5
+    """This is a Generation 2 shim for a Generation3 Butler.
+    """
+
+    def __init__(self, butlerSubset, dataId):
```

Comment:
Why are we taking a subset here, it seems like it is not really used in lue of the datasetType. Is it just because of the API of the old DataRef?

Reply:
It's just the old API - we almost always use `DataRef` as a combination of a dataset type and a data ID, but it originated as the thing iterated over by a `ButlerSubset`.  One relic of that is that it's moderately common to see `dataRef.butlerSubset.butler`, and I wanted to support that.